src/pudl/extract/csv.py

```diff
@@ -0,0 +1,106 @@
+"""Extractor for CSV data."""
+from csv import DictReader
+from importlib import resources
+
+import pandas as pd
+from dagster import AssetsDefinition, OpDefinition, graph_asset, op
+
+import pudl.logging_helpers
+from pudl.workspace.datastore import Datastore
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+class CsvExtractor:
+    """Generalized class for extracting dataframes from CSV files.
```

Comment:
So this CsvExtractor class requires subclassing to set the `DATASET` - that seems a little clunky. If we can pass the dataset name to `__init__`, we'll have a `CsvExtractor` class which exposes:

```python
extractor = CsvExtractor(datastore, dataset_name)
extractor.read_source(filename) # get one file
extractor.extract() # get all files
```

Which seems like a nice generic CSV extraction interface that doesn't require subclassing to read a variety of different collections of CSV files.

I think we might want to tweak that API a bit to expose table-level operations:

```python
extractor.extract_one(table_name) # pd.DataFrame
extractor.extract_all() # dict[str, pd.DataFrame]
```

Because then we can read multiple tables' files in, in parallel - each table could have its own asset like

```python
@asset 
def raw_eia176__table_name(context):
    ...
    extractor.extract_one(table_name)
```

Which is pretty straightforward to factory-ize if you want to make a bunch of these assets.

In the event we need to read & combine multiple files for a single table (like we see in EIA 860), we can turn that simple asset above into a graph-backed asset. But for EIA176 that seems like overkill.

Reply:
I think there's also an argument for passing in the table -> file(s) map and the path to file as init params, and then using the datastore at the call site - this lets people use the extractor to explore data sets that we haven't actually integrated into our upstream work yet, and lets you pass stuff in for testing without extensive patching. Could be something like

```
class CsvExtractor:
    def __init__(self, path: pathlib.Path | zipfile.Path, table_files_map: dict[str, list[str]]):
        ...
    @classmethod
    def from_resource(cls, datastore: Datastore, resource_id: str):
        # existing logic to get zipfile + table/file map
        return cls(...)
```

What do you think?

Reply:
> So this CsvExtractor class requires subclassing to set the DATASET - that seems a little clunky. If we can pass the dataset name to __init__

Meant to comment on this. I actually started using dataset as a construction parameter but stopped because I was introducing a competing pattern, e.g., vs `FercDbfExtractor`'s inheritance tree. I'll proceed with parameterizing over inheritance now that someone else is also inclined.

> I think there's also an argument for passing in the table -> file(s) map and the path to file as init params, and then using the datastore at the call site

I don't know what the datastore would be used for in that case, since right now it's only to get the zipfile path. But generally you're talking about providing a class that lets a user/client point to any zip file and get dataframes out of it based on a `table_files_map`? Right now the zipfile path and the table-file(s) map are coupled on the dataset name, e.g., `eia176`. Decoupling opens a wider window for invalid combos, i.e., table filenames that do not exist in the zip archive, but yeah, it would be nice to be able to develop against data without needing the source published on Zenodo. I'll take a pass in that direction.